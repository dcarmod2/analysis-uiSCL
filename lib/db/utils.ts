import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollection(obj) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollection)
  if (typeof obj === 'object') return mapValues(obj, serializeCollection)
  return obj
}
