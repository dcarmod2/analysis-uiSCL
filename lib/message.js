import get from 'lodash/get'

import messages from '../i18n/en.json'

/**
 * Expose a Set of all the keys used
 */
export const KeysUsed = new Set()

/**
 * Requires a key, defaultMessage and parameters are optional
 */
export default function getMessage(key, defaultMessage, parameters) {
  if (defaultMessage == null) {
    defaultMessage = ''
    parameters = {}
  } else if (typeof defaultMessage === 'object') {
    parameters = defaultMessage
    defaultMessage = ''
  }

  // Store the used key
  KeysUsed.add(key)

  // Get the message with "lodash/get" to allow nested keys ('noun.action' => {noun: {action: 'value'}})
  const msg = get(messages, key, defaultMessage)
  const result = parameters ? replaceMessage(msg, parameters) : msg
  return result
}

function replaceMessage(msg, data) {
  return msg.replace(
    new RegExp('%\\((' + Object.keys(data).join('|') + ')\\)', 'g'),
    (m, key) =>
      Object.prototype.hasOwnProperty.call(data, key) ? data[key] : m
  )
}
