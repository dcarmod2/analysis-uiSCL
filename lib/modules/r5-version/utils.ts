import {MINIMUM_R5_VERSION, VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version: string): number {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || '0') / 100
    )
  }

  return 0
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = versionToNumber(version) % 1 !== 0
  return isTestVersion || workerVersionInRange(version, min, max)
}
