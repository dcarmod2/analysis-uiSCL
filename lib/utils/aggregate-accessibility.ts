import get from 'lodash/get'
import range from 'lodash/range'

const N_HISTOGRAM_BINS = 15

/**
 * Compute the accessibility, exported for use in comparison weighted average
 * accessibility.
 * @param weights a grid containing the weights (e.g. population)
 * @param aggregationArea a grid containing the aggregation area, with the
 *        percentage the aggregation area overlaps each cell, scaled between 0
 *        (no overlap) and 100,000 (full coverage)
 * @param accessibility a grid containing the accessibility (i.e. the regional
 *        analysis results)
 */
export default function getAggregateAccessibility(
  accessibility,
  aggregationArea,
  weights
) {
  const aa = get(aggregationArea, 'grid')
  if (weights == null || aa == null || accessibility == null) {
    return
  }

  if (accessibility.zoom !== weights.zoom || accessibility.zoom !== aa.zoom) {
    console.error('Zoom levels of all grids in weighted average must match!')
  }

  // find max and min value for histogram
  let maxAccessibility = 0
  let minAccessibility = Infinity

  // Pass one: find min and max from aggregationArea
  // aggregationArea is probably the smallest grid and is guaranteed to contain
  // all pixels
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    // Transform to coordinates in other grid; all grids have same zoom, so
    // translate based on the difference in where the edges of the grid are
    const accessY = y + aa.north - accessibility.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        minAccessibility = Math.min(accessibilityThisCell, minAccessibility)
        maxAccessibility = Math.max(accessibilityThisCell, maxAccessibility)
      }
    }
  }

  const spread = maxAccessibility - minAccessibility
  const bins = range(N_HISTOGRAM_BINS).map(() => 0)

  let sumOfWeights = 0
  // weight and accessibility value for each cell, used to compute percentiles
  const weightedAccessibility = []

  // pass two: accumulate into bins
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    const accessY = y + aa.north - accessibility.north
    const weightY = y + aa.north - weights.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        let binIndex = Math.floor(
          ((accessibilityThisCell - minAccessibility) / spread) *
            N_HISTOGRAM_BINS
        )

        if (binIndex > N_HISTOGRAM_BINS) console.error('invalid bin!')

        // avoid putting the highest accessibility cell beyond the histogram
        if (binIndex === N_HISTOGRAM_BINS) binIndex--

        // how much of this cell is within the aggregationArea
        const aggregationAreaWeightThisCell = aa.data[pixel] / 100000

        const weightX = x + aa.west - weights.west
        const weightThisCell = weights.contains(weightX, weightY)
          ? weights.data[weightY * weights.width + weightX]
          : 0

        const weight = aggregationAreaWeightThisCell * weightThisCell
        bins[binIndex] += aggregationAreaWeightThisCell * weightThisCell

        sumOfWeights += weight
        weightedAccessibility.push({
          weight,
          accessibility: accessibilityThisCell
        })
      }
    }
  }

  // Figure out percentiles
  let cumulativeWeight = 0
  // Preinitialize with null so 50th percentile is percentiles[50].
  // There is no zeroth percentile
  const percentiles = [0]
  const weightPerPercentile = sumOfWeights / 100
  let expectedWeight = weightPerPercentile
  weightedAccessibility.sort((a, b) => a.accessibility - b.accessibility)

  // we calculate and return an old-style weighted average for comparison purposes.
  let sumOfWeightedAccessibility = 0

  // We loop over the sorted weighted accessibility values. When we cross a
  // percentile boundary, we push that percentile onto the list
  weightedAccessibility.forEach(({accessibility, weight}) => {
    sumOfWeightedAccessibility += accessibility * weight

    cumulativeWeight += weight

    // while not if, possible to cross many percentiles with one high-density cell
    while (cumulativeWeight > expectedWeight) {
      // There is a percentile value somewhere within the values at this sample
      // We use > above, not >=, so that a percentile value that falls exactly
      // between two accessibility values will get the higher value, e.g. if 10
      // people have accessibility 50 and ten people have accessibility 60, we
      // want 60. This is because in the UI we say "x% of <people> have
      // accessibility y or higher," so we want y to be as high as possible
      // while still being a true statement.
      percentiles.push(accessibility)
      expectedWeight += weightPerPercentile
    }
  })

  // includes values 1 - 99
  // When the resolution of the data is low the last item may only define (say)
  // the 90th percentile, all remaining percentiles are just the max. Also
  // handles the case when there is no accessibility within the aggregation
  // area by returning all zeros.
  while (percentiles.length < 99) {
    percentiles.push(
      weightedAccessibility.length > 0
        ? weightedAccessibility[weightedAccessibility.length - 1].accessibility
        : 0
    )
  }

  const binWidth = spread / N_HISTOGRAM_BINS

  return {
    minAccessibility,
    maxAccessibility,
    percentiles,
    weightedAverage: sumOfWeightedAccessibility / sumOfWeights,
    bins: bins.map((value, idx) => ({
      value,
      min: minAccessibility + idx * binWidth,
      max: minAccessibility + (idx + 1) * binWidth
    }))
  }
}
