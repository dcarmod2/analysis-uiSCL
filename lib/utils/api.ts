// Query values are typed as string OR string[]
export function getQueryAsString(p: string | string[]): string {
  return Array.isArray(p) ? p[0] : p
}

// Get a query as an object
export function getQueryAsObject(
  p: string | string[]
): Record<string, unknown> {
  try {
    return JSON.parse(getQueryAsString(p))
  } catch (e) {
    return {}
  }
}

// Convert an error to POJO for sending over the wire
export function errorToPOJO(e: Error): Record<string, unknown> {
  const obj = {}
  Object.getOwnPropertyNames(e).forEach((k) => {
    obj[k] = e[k]
  })
  return obj
}
