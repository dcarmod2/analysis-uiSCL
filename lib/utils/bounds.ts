import lonlat from '@conveyal/lonlat'

import L from 'lib/leaflet'
import message from 'lib/message'

import reprojectCoordinates from './reproject-coordinates'

const ANTIMERIDIAN_WARNING =
  'We do not currently support bounds that cross the antimeridian.'

type Bounds = {
  north: number
  south: number
  east: number
  west: number
  toString: () => string
}

export function toLatLngBounds(
  bounds: Bounds | L.LatLngBounds
): L.LatLngBounds {
  // Check if it's already a leaflet bounds object
  if (bounds instanceof L.LatLngBounds) return bounds
  return new L.LatLngBounds([
    [bounds.north, bounds.west],
    [bounds.south, bounds.east]
  ])
}

/**
 * NB: We do not currently support crossing the antimeridian.
 */
export function fromLatLngBounds(bounds: L.LatLngBounds) {
  // Don't change the original bounds object
  let b = new L.LatLngBounds(bounds.getSouthWest(), bounds.getNorthEast())

  if (b.getWest() < -180) {
    b = new L.LatLngBounds(new L.LatLng(b.getSouth(), -180), b.getNorthEast())
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  if (b.getEast() > 180) {
    b = new L.LatLngBounds(new L.LatLng(b.getSouth(), 180), b.getNorthEast())
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  const bboxString = b.toBBoxString()
  return {
    north: b.getNorth(),
    south: b.getSouth(),
    east: b.getEast(),
    west: b.getWest(),
    toString: () => bboxString
  }
}

/**
 * Reproject bounds to the grid size that we use.
 */
export function reprojectBounds(bounds) {
  return new L.LatLngBounds(
    lonlat.toLeaflet(reprojectCoordinates(bounds.getSouthWest())),
    lonlat.toLeaflet(reprojectCoordinates(bounds.getNorthEast()))
  )
}
