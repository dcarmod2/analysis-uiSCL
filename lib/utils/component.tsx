import {ThemeProvider} from '@chakra-ui/core'
import {expect} from '@jest/globals'
import enzyme from 'enzyme'
import {Router} from 'next/router'
import {RouterContext} from 'next/dist/next-server/lib/router-context'
import React from 'react'
import {Map} from 'react-leaflet'
import {Provider} from 'react-redux'

import {makeMockStore, mockStores} from './mock-data'

// Next.js uses Math.random to generate unique ids for pages that get captured in snapshots.
global.Math.random = () => 0.5

// Set __NEXT_DATA__ (required by Router)
window.__NEXT_DATA__ = {} as any
const router = new Router('page', {}, 'as-path', {} as any)

const getName = (c) => c.displayName || c.name

export function testComponent(
  TestComponent,
  props = {},
  initialData = mockStores.init
) {
  const store = makeMockStore(initialData)
  const Wrapping = (props) => (
    <Provider store={store}>
      <ThemeProvider>
        <RouterContext.Provider value={router}>
          {props.children}
        </RouterContext.Provider>
      </ThemeProvider>
    </Provider>
  )

  const mount = () =>
    enzyme.mount(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // TODO: Dive down through the `TestComponent` automatically
  const shallow = () =>
    enzyme.shallow(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // For debugging purposes (this is not the original component!)
  const name = `WithContext(${getName(TestComponent)})`

  return {mount, name, router, shallow, store}
}

/**
 * Rename to `wrapComponent`. TODO replace all `testComponent`
 */
export const wrapComponent = testComponent

/**
 * Helper function to create quick snapshot tests for components.
 */
export function testAndSnapshot(C, p, id?: any) {
  test(`Component ${getName(C)} snapshot(mount)`, () => {
    const c = testComponent(C, p, id)
    const t = c.mount()
    expect(t).toMatchSnapshot()
    t.unmount()
  })
}

/**
 * Test and snapshot map component
 */
export function wrapMapComponent(C, p, id?: any) {
  const MapWrapper = (props) => (
    <Map>
      <C {...props} />
    </Map>
  )
  MapWrapper.displayName = `WithMap(${getName(C)})`
  return testComponent(MapWrapper, p, id)
}

function snapshotTest(wc, type) {
  test(`Component ${getName(wc)} snapshot(${type})`, () => {
    const s = wc[`${type}`]()
    expect(s).toMatchSnapshot()
    s.unmount()
  })
}

/**
 * Shallow snapshot test.
 */
export function shallowSnapshot(wc) {
  snapshotTest(wc, 'shallow')
}

/**
 * Deep snapshot test.
 */
export function deepSnapshot(wc) {
  snapshotTest(wc, 'mount')
}
