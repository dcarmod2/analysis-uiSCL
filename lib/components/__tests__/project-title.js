import message from 'lib/message'
import {mockProject} from 'lib/utils/mock-data'
import {testComponent} from 'lib/utils/component'

import ProjectTitle from '../project-title'

test('Component > ProjectTitle', () => {
  const p = testComponent(ProjectTitle, {project: mockProject})
  const tree = p.mount()
  expect(tree).toMatchSnapshot()

  // Modal should not exist yet
  expect(tree.exists('ExportProject')).toBe(false)

  // Show export dialog
  tree
    .find(`button[aria-label="${message('project.export')}"]`)
    .simulate('click')

  // find contents of modal
  expect(tree.exists('ExportProject')).toBe(true)

  // Unmount without errors
  tree.unmount()
})
