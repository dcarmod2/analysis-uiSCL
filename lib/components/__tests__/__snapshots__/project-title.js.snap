// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component > ProjectTitle 1`] = `
<ProjectTitle
  project={
    Object {
      "_id": "1",
      "accessGroup": "accessGroupName",
      "bundleId": "1",
      "createdAt": "2017-11-08T12:30:50.133Z",
      "createdBy": "person@gmail.com",
      "name": "Mock Project",
      "nonce": "12345",
      "regionId": "1",
      "updatedAt": "2017-11-08T12:31:19.602Z",
      "updatedBy": "person@gmail.com",
      "variants": Array [
        "Default",
      ],
    }
  }
>
  <Flex
    align="center"
    borderBottom="1px solid #E2E8F0"
    p={2}
    width="320px"
  >
    <Box
      alignItems="center"
      borderBottom="1px solid #E2E8F0"
      display="flex"
      p={2}
      width="320px"
    >
      <div
        className="css-hyly4r"
      >
        <IconButton
          icon={
            Object {
              "icon": Array [],
              "iconName": "chevron-left",
              "prefix": "fas",
            }
          }
          label="All projects"
          onClick={[Function]}
        >
          <Box>
            <div
              className="css-0"
            >
              <Tip
                label="All projects"
                placement="auto"
              >
                <Tooltip
                  aria-label="Tooltip: All projects"
                  fontSize="md"
                  hasArrow={true}
                  label="All projects"
                  placement="auto"
                  zIndex={1500}
                >
                  <Button
                    aria-label="All projects"
                    isActive={false}
                    isDisabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    size="sm"
                    variant="ghost"
                    variantColor="blue"
                  >
                    <PseudoBox
                      _active={
                        Object {
                          "bg": "blue.100",
                        }
                      }
                      _disabled={
                        Object {
                          "boxShadow": "none",
                          "cursor": "not-allowed",
                          "opacity": "40%",
                        }
                      }
                      _focus={
                        Object {
                          "boxShadow": "outline",
                        }
                      }
                      _hover={
                        Object {
                          "bg": "blue.50",
                        }
                      }
                      alignItems="center"
                      appearance="none"
                      aria-label="All projects"
                      as="button"
                      bg="transparent"
                      borderRadius="md"
                      color="blue.500"
                      display="inline-flex"
                      fontSize="sm"
                      fontWeight="semibold"
                      height={8}
                      justifyContent="center"
                      lineHeight="1.2"
                      minWidth={8}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      outline="none"
                      position="relative"
                      px={3}
                      transition="all 250ms"
                      type="button"
                      userSelect="none"
                      verticalAlign="middle"
                      whiteSpace="nowrap"
                    >
                      <button
                        aria-label="All projects"
                        className="css-ig2uvu"
                        onBlur={[Function]}
                        onClick={[Function]}
                        onFocus={[Function]}
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        type="button"
                      >
                        <Icon
                          fixedWidth={false}
                          icon={
                            Object {
                              "icon": Array [],
                              "iconName": "chevron-left",
                              "prefix": "fas",
                            }
                          }
                        >
                          <FontAwesomeIcon
                            border={false}
                            className=""
                            fixedWidth={false}
                            flip={null}
                            icon={
                              Object {
                                "icon": Array [],
                                "iconName": "chevron-left",
                                "prefix": "fas",
                              }
                            }
                            inverse={false}
                            listItem={false}
                            mask={null}
                            pull={null}
                            pulse={false}
                            rotation={null}
                            size={null}
                            spin={false}
                            swapOpacity={false}
                            symbol={false}
                            title=""
                            transform={null}
                          >
                            <svg
                              aria-hidden="true"
                              className="svg-inline--fa fa-chevron-left fa-w-NaN "
                              data-icon="chevron-left"
                              data-prefix="fas"
                              focusable="false"
                              role="img"
                              style={Object {}}
                              viewBox="0 0 undefined undefined"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                fill="currentColor"
                                style={Object {}}
                              />
                            </svg>
                          </FontAwesomeIcon>
                        </Icon>
                      </button>
                    </PseudoBox>
                  </Button>
                  <ForwardRef
                    anchorEl={
                      <button
                        aria-label="All projects"
                        class="css-ig2uvu"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="svg-inline--fa fa-chevron-left fa-w-NaN "
                          data-icon="chevron-left"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          viewBox="0 0 undefined undefined"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            fill="currentColor"
                          />
                        </svg>
                      </button>
                    }
                    arrowSize="10px"
                    bg="gray.700"
                    borderRadius="sm"
                    color="whiteAlpha.900"
                    fontSize="md"
                    fontWeight="medium"
                    hasArrow={true}
                    isOpen={false}
                    maxW="320px"
                    modifiers={
                      Object {
                        "offset": Object {
                          "enabled": true,
                          "offset": "0, 8",
                        },
                      }
                    }
                    placement="auto"
                    pointerEvents="none"
                    px="8px"
                    py="2px"
                    shadow="md"
                    timeout={200}
                    usePortal={true}
                    zIndex={1500}
                  />
                </Tooltip>
              </Tip>
            </div>
          </Box>
        </IconButton>
        <Heading
          flex="1"
          ml={2}
          overflow="hidden"
          size="md"
          style={
            Object {
              "textOverflow": "ellipsis",
            }
          }
          whiteSpace="nowrap"
        >
          <Box
            as="h2"
            flex="1"
            fontFamily="heading"
            fontSize="xl"
            fontWeight="bold"
            lineHeight="shorter"
            ml={2}
            overflow="hidden"
            style={
              Object {
                "textOverflow": "ellipsis",
              }
            }
            whiteSpace="nowrap"
          >
            <h2
              className="css-1ruodvp"
              style={
                Object {
                  "textOverflow": "ellipsis",
                }
              }
            >
              Mock Project
            </h2>
          </Box>
        </Heading>
        <Flex>
          <Box
            display="flex"
          >
            <div
              className="css-k008qs"
            >
              <IconButton
                icon={
                  Object {
                    "icon": Array [],
                    "iconName": "external-link-alt",
                    "prefix": "fas",
                  }
                }
                label="Download or share this project"
                onClick={[Function]}
              >
                <Box>
                  <div
                    className="css-0"
                  >
                    <Tip
                      label="Download or share this project"
                      placement="auto"
                    >
                      <Tooltip
                        aria-label="Tooltip: Download or share this project"
                        fontSize="md"
                        hasArrow={true}
                        label="Download or share this project"
                        placement="auto"
                        zIndex={1500}
                      >
                        <Button
                          aria-label="Download or share this project"
                          isActive={false}
                          isDisabled={false}
                          onBlur={[Function]}
                          onClick={[Function]}
                          onFocus={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          size="sm"
                          variant="ghost"
                          variantColor="blue"
                        >
                          <PseudoBox
                            _active={
                              Object {
                                "bg": "blue.100",
                              }
                            }
                            _disabled={
                              Object {
                                "boxShadow": "none",
                                "cursor": "not-allowed",
                                "opacity": "40%",
                              }
                            }
                            _focus={
                              Object {
                                "boxShadow": "outline",
                              }
                            }
                            _hover={
                              Object {
                                "bg": "blue.50",
                              }
                            }
                            alignItems="center"
                            appearance="none"
                            aria-label="Download or share this project"
                            as="button"
                            bg="transparent"
                            borderRadius="md"
                            color="blue.500"
                            display="inline-flex"
                            fontSize="sm"
                            fontWeight="semibold"
                            height={8}
                            justifyContent="center"
                            lineHeight="1.2"
                            minWidth={8}
                            onBlur={[Function]}
                            onClick={[Function]}
                            onFocus={[Function]}
                            onMouseEnter={[Function]}
                            onMouseLeave={[Function]}
                            outline="none"
                            position="relative"
                            px={3}
                            transition="all 250ms"
                            type="button"
                            userSelect="none"
                            verticalAlign="middle"
                            whiteSpace="nowrap"
                          >
                            <button
                              aria-label="Download or share this project"
                              className="css-ig2uvu"
                              onBlur={[Function]}
                              onClick={[Function]}
                              onFocus={[Function]}
                              onMouseEnter={[Function]}
                              onMouseLeave={[Function]}
                              type="button"
                            >
                              <Icon
                                fixedWidth={false}
                                icon={
                                  Object {
                                    "icon": Array [],
                                    "iconName": "external-link-alt",
                                    "prefix": "fas",
                                  }
                                }
                              >
                                <FontAwesomeIcon
                                  border={false}
                                  className=""
                                  fixedWidth={false}
                                  flip={null}
                                  icon={
                                    Object {
                                      "icon": Array [],
                                      "iconName": "external-link-alt",
                                      "prefix": "fas",
                                    }
                                  }
                                  inverse={false}
                                  listItem={false}
                                  mask={null}
                                  pull={null}
                                  pulse={false}
                                  rotation={null}
                                  size={null}
                                  spin={false}
                                  swapOpacity={false}
                                  symbol={false}
                                  title=""
                                  transform={null}
                                >
                                  <svg
                                    aria-hidden="true"
                                    className="svg-inline--fa fa-external-link-alt fa-w-NaN "
                                    data-icon="external-link-alt"
                                    data-prefix="fas"
                                    focusable="false"
                                    role="img"
                                    style={Object {}}
                                    viewBox="0 0 undefined undefined"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      fill="currentColor"
                                      style={Object {}}
                                    />
                                  </svg>
                                </FontAwesomeIcon>
                              </Icon>
                            </button>
                          </PseudoBox>
                        </Button>
                        <ForwardRef
                          anchorEl={
                            <button
                              aria-label="Download or share this project"
                              class="css-ig2uvu"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="svg-inline--fa fa-external-link-alt fa-w-NaN "
                                data-icon="external-link-alt"
                                data-prefix="fas"
                                focusable="false"
                                role="img"
                                viewBox="0 0 undefined undefined"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  fill="currentColor"
                                />
                              </svg>
                            </button>
                          }
                          arrowSize="10px"
                          bg="gray.700"
                          borderRadius="sm"
                          color="whiteAlpha.900"
                          fontSize="md"
                          fontWeight="medium"
                          hasArrow={true}
                          isOpen={false}
                          maxW="320px"
                          modifiers={
                            Object {
                              "offset": Object {
                                "enabled": true,
                                "offset": "0, 8",
                              },
                            }
                          }
                          placement="auto"
                          pointerEvents="none"
                          px="8px"
                          py="2px"
                          shadow="md"
                          timeout={200}
                          usePortal={true}
                          zIndex={1500}
                        />
                      </Tooltip>
                    </Tip>
                  </div>
                </Box>
              </IconButton>
              <IconButton
                icon={
                  Object {
                    "icon": Array [],
                    "iconName": "cog",
                    "prefix": "fas",
                  }
                }
                label="Edit project settings"
                onClick={[Function]}
              >
                <Box>
                  <div
                    className="css-0"
                  >
                    <Tip
                      label="Edit project settings"
                      placement="auto"
                    >
                      <Tooltip
                        aria-label="Tooltip: Edit project settings"
                        fontSize="md"
                        hasArrow={true}
                        label="Edit project settings"
                        placement="auto"
                        zIndex={1500}
                      >
                        <Button
                          aria-label="Edit project settings"
                          isActive={false}
                          isDisabled={false}
                          onBlur={[Function]}
                          onClick={[Function]}
                          onFocus={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          size="sm"
                          variant="ghost"
                          variantColor="blue"
                        >
                          <PseudoBox
                            _active={
                              Object {
                                "bg": "blue.100",
                              }
                            }
                            _disabled={
                              Object {
                                "boxShadow": "none",
                                "cursor": "not-allowed",
                                "opacity": "40%",
                              }
                            }
                            _focus={
                              Object {
                                "boxShadow": "outline",
                              }
                            }
                            _hover={
                              Object {
                                "bg": "blue.50",
                              }
                            }
                            alignItems="center"
                            appearance="none"
                            aria-label="Edit project settings"
                            as="button"
                            bg="transparent"
                            borderRadius="md"
                            color="blue.500"
                            display="inline-flex"
                            fontSize="sm"
                            fontWeight="semibold"
                            height={8}
                            justifyContent="center"
                            lineHeight="1.2"
                            minWidth={8}
                            onBlur={[Function]}
                            onClick={[Function]}
                            onFocus={[Function]}
                            onMouseEnter={[Function]}
                            onMouseLeave={[Function]}
                            outline="none"
                            position="relative"
                            px={3}
                            transition="all 250ms"
                            type="button"
                            userSelect="none"
                            verticalAlign="middle"
                            whiteSpace="nowrap"
                          >
                            <button
                              aria-label="Edit project settings"
                              className="css-ig2uvu"
                              onBlur={[Function]}
                              onClick={[Function]}
                              onFocus={[Function]}
                              onMouseEnter={[Function]}
                              onMouseLeave={[Function]}
                              type="button"
                            >
                              <Icon
                                fixedWidth={false}
                                icon={
                                  Object {
                                    "icon": Array [],
                                    "iconName": "cog",
                                    "prefix": "fas",
                                  }
                                }
                              >
                                <FontAwesomeIcon
                                  border={false}
                                  className=""
                                  fixedWidth={false}
                                  flip={null}
                                  icon={
                                    Object {
                                      "icon": Array [],
                                      "iconName": "cog",
                                      "prefix": "fas",
                                    }
                                  }
                                  inverse={false}
                                  listItem={false}
                                  mask={null}
                                  pull={null}
                                  pulse={false}
                                  rotation={null}
                                  size={null}
                                  spin={false}
                                  swapOpacity={false}
                                  symbol={false}
                                  title=""
                                  transform={null}
                                >
                                  <svg
                                    aria-hidden="true"
                                    className="svg-inline--fa fa-cog fa-w-NaN "
                                    data-icon="cog"
                                    data-prefix="fas"
                                    focusable="false"
                                    role="img"
                                    style={Object {}}
                                    viewBox="0 0 undefined undefined"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      fill="currentColor"
                                      style={Object {}}
                                    />
                                  </svg>
                                </FontAwesomeIcon>
                              </Icon>
                            </button>
                          </PseudoBox>
                        </Button>
                        <ForwardRef
                          anchorEl={
                            <button
                              aria-label="Edit project settings"
                              class="css-ig2uvu"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="svg-inline--fa fa-cog fa-w-NaN "
                                data-icon="cog"
                                data-prefix="fas"
                                focusable="false"
                                role="img"
                                viewBox="0 0 undefined undefined"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  fill="currentColor"
                                />
                              </svg>
                            </button>
                          }
                          arrowSize="10px"
                          bg="gray.700"
                          borderRadius="sm"
                          color="whiteAlpha.900"
                          fontSize="md"
                          fontWeight="medium"
                          hasArrow={true}
                          isOpen={false}
                          maxW="320px"
                          modifiers={
                            Object {
                              "offset": Object {
                                "enabled": true,
                                "offset": "0, 8",
                              },
                            }
                          }
                          placement="auto"
                          pointerEvents="none"
                          px="8px"
                          py="2px"
                          shadow="md"
                          timeout={200}
                          usePortal={true}
                          zIndex={1500}
                        />
                      </Tooltip>
                    </Tip>
                  </div>
                </Box>
              </IconButton>
            </div>
          </Box>
        </Flex>
      </div>
    </Box>
  </Flex>
</ProjectTitle>
`;
